### DISCUSSION TITLE
Restricted Functions for Use with Transaction Mode

### DISCUSSION THREAD

**Discussion by El-Tommy** [pgbouncer/pgbouncer#651] [Q&A]
I've always used PgBouncer in session mode, but for my use case the performance isn't so good, as PHP scripts that run for a long time end up with a long session connected to the database (even for minutes), without releasing the connection for the other requests. Increased the number of connections in IDLE until reaching the limit of connections.

I tried switching to Transaction mode, but for lack of knowledge, it turned out to be a bad idea.

As I found out the hard way, some PostgreSQL functions require to be in the same session as the previous command, such as lastval(), currval() after an insert.

Reading the PgBouncer [documentation](https://www.pgbouncer.org/features.html) there is a list of functions that are not accepted in transaction mode, but I haven't found any documentation (even PostgreSQL) that says all functions that must run in the same session after an operation (insert, delete, update, etc. ).

Is there any location with all these restrictions documented?

> Note: As I use the CodeIgniter framework for PHP, the insert command is executed separately from lastval(), 
> currval(). That's why I can't execute both in the same transaction with the database.

**Comment by eulerto** [comment]
You probably misunderstood how transaction mode works. There is no guarantee that a client will use the same backend connection in 2 subsequent commands. 2 subsequent commands without a transaction indicator are 2 separate (implicit) transactions. Having said that, the second transaction (probably using a different backend connection) won't have run the `nextval` function so you will get an error that the sequence is not yet defined in the session.

[Postgres documentation](https://www.postgresql.org/docs/current/functions-sequence.html) mentions the fact that `currval` and `lastval` should be called after `nextval` in the **same** session.

> **currval ( regclass ) → bigint**
> Returns the value most recently obtained by nextval for this sequence in the current session. (An error is reported if nextval has never been called for this sequence in this session.) Because this is returning a session-local value, it gives a predictable answer whether or not other sessions have executed nextval since the current session did.

> **lastval () → bigint**
> Returns the value most recently returned by nextval in the current session. This function is identical to currval, except that instead of taking the sequence name as an argument it refers to whichever sequence nextval was most recently applied to in the current session. It is an error to call lastval if nextval has not yet been called in the current session.

If your framework does not take this into consideration, it should be fixed (a transaction that covers your command and currval/lastval) or even consider using the RETURNING clause.

**Reply by El-Tommy** [reply]
@eulerto I'm using CodeIgniter, in their documentation I found trans_begin and trans_complete that start a transaction sending BEGIN and when it finishes executing the SQL commands it sends a COMMIT command.
In my tests, between the SQL commands I put a sleep of a few seconds and in all executed commands the backend_xid is the same.
But due to my traumas of using another session ID I'm considering using the RETURNING command as it already does everything in a single request to the server.

https://codeigniter.com/userguide3/database/transactions.html
